astropy/convolution/convolve.py

```diff
@@ -137,66 +137,83 @@ def convolve(array, kernel, boundary='fill', fill_value=0.,
     # but, if we just so happen to be lucky enough to have the input array
     # have exactly the desired type, we just alias to array_internal
 
-    # Check if kernel is kernel instance
-    if isinstance(kernel, Kernel):
-        # Check if array is also kernel instance, if so convolve and
-        # return new kernel instance
-        if isinstance(array, Kernel):
-            if isinstance(array, Kernel1D) and isinstance(kernel, Kernel1D):
-                new_array = convolve1d_boundary_fill(array.array, kernel.array,
-                                                     0, True)
-                new_kernel = Kernel1D(array=new_array)
-            elif isinstance(array, Kernel2D) and isinstance(kernel, Kernel2D):
-                new_array = convolve2d_boundary_fill(array.array, kernel.array,
-                                                     0, True)
-                new_kernel = Kernel2D(array=new_array)
-            else:
-                raise Exception("Can't convolve 1D and 2D kernel.")
-            new_kernel._separable = kernel._separable and array._separable
-            new_kernel._is_bool = False
-            return new_kernel
-        kernel = kernel.array
+    # If array and kernel are both Kernal instances, convolve here and return.
+    if isinstance(kernel, Kernel) and isinstance(array, Kernel):
+        if isinstance(array, Kernel1D) and isinstance(kernel, Kernel1D):
+            new_array = convolve1d_boundary_fill(array.array, kernel.array,
+                                                 0, True)
+            new_kernel = Kernel1D(array=new_array)
+        elif isinstance(array, Kernel2D) and isinstance(kernel, Kernel2D):
+            new_array = convolve2d_boundary_fill(array.array, kernel.array,
+                                                 0, True)
+            new_kernel = Kernel2D(array=new_array)
+        else:
+            raise Exception("Can't convolve 1D and 2D kernel.")
+        new_kernel._separable = kernel._separable and array._separable
+        new_kernel._is_bool = False
```

Comment:
can you comment on what these hidden variables are for?

Reply:
Will do, but FYI this code already existed.